### ISSUE TITLE
Error message - Connection Draining

### ISSUE THREAD

**Issue by roelvanloon** [mendix/MxPushNotifications#45]
I have implemented Pushnotifications version 2.1.0 for Mx 6.8.1. Recently we have changed the Google project from GCM to FCM (firebase). Pushnotifications have working fine. Since a few days we receive error messages regarding connection draining. Restarting GCM does not work, new Google project with new api key does also not work. Can you help us out?

Find the error log below.

Thanks.

```
Oct 18 21:28:10.019 127.0.0.1 tr10000: INFO - PushNotifications: GCM: Dropping downstream message since the connection is draining
Oct 18 21:28:10.020 127.0.0.1 tr10000: ERROR - PushNotifications: GCM: Message to d8eqJvAo... failed: java.lang.Exception: Message not delivered, view log.
Oct 18 21:28:10.021 127.0.0.1 tr10000: ERROR - PushNotifications: (1/38) java.lang.Exception: Message not delivered, view log.
Oct 18 21:28:10.021 127.0.0.1 tr10000: ERROR - PushNotifications: (2/38)    at pushnotifications.implementation.gcm.GCMConnection.sendMessage(GCMConnection.java:337)
Oct 18 21:28:10.021 127.0.0.1 tr10000: ERROR - PushNotifications: (3/38)    at pushnotifications.implementation.gcm.GCMConnection$$Lambda$13/887504894.accept(Unknown Source)
Oct 18 21:28:10.021 127.0.0.1 tr10000: ERROR - PushNotifications: (4/38)    at java.util.ArrayList.forEach(ArrayList.java:1249)
Oct 18 21:28:10.021 127.0.0.1 tr10000: ERROR - PushNotifications: (5/38)    at pushnotifications.implementation.gcm.GCMConnection.sendMessages(GCMConnection.java:129)
Oct 18 21:28:10.021 127.0.0.1 tr10000: ERROR - PushNotifications: (6/38)    at pushnotifications.actions.SendGoogleMessages.executeAction(SendGoogleMessages.java:38)
Oct 18 21:28:10.021 127.0.0.1 tr10000: ERROR - PushNotifications: (7/38)    at pushnotifications.actions.SendGoogleMessages.executeAction(SendGoogleMessages.java:17)
Oct 18 21:28:10.021 127.0.0.1 tr10000: ERROR - PushNotifications: (8/38)    at com.mendix.systemwideinterfaces.core.UserAction.execute(UserAction.java:49)
Oct 18 21:28:10.021 127.0.0.1 tr10000: ERROR - PushNotifications: (9/38)    at com.mendix.core.actionmanagement.CoreAction.doCall(CoreAction.java:260)
Oct 18 21:28:10.021 127.0.0.1 tr10000: ERROR - PushNotifications: (10/38)   at com.mendix.core.actionmanagement.CoreAction.call(CoreAction.java:248)
Oct 18 21:28:10.021 127.0.0.1 tr10000: ERROR - PushNotifications: (11/38)   at com.mendix.core.actionmanagement.ActionManager$1.execute(ActionManager.java:168)
Oct 18 21:28:10.021 127.0.0.1 tr10000: ERROR - PushNotifications: (12/38)   at com.mendix.util.classloading.Runner.doRunUsingClassLoaderOf(Runner.java:36)
Oct 18 21:28:10.021 127.0.0.1 tr10000: ERROR - PushNotifications: (13/38)   at com.mendix.core.actionmanagement.ActionManager.executeSync(ActionManager.java:174)
Oct 18 21:28:10.021 127.0.0.1 tr10000: ERROR - PushNotifications: (14/38)   at com.mendix.core.component.InternalCore.execute(InternalCore.java:390)
Oct 18 21:28:10.022 127.0.0.1 tr10000: ERROR - PushNotifications: (15/38)   at com.mendix.modules.microflowengine.actions.actioncall.JavaAction.execute(JavaAction.scala:52)
Oct 18 21:28:10.022 127.0.0.1 tr10000: ERROR - PushNotifications: (16/38)   at com.mendix.modules.microflowengine.microflow.impl.MicroflowObject.execute(MicroflowObject.java:47)
Oct 18 21:28:10.022 127.0.0.1 tr10000: ERROR - PushNotifications: (17/38)   at com.mendix.modules.microflowengine.microflow.impl.MicroflowImpl.executeAfterBreakingIfNecessary(MicroflowImpl.java:192)
Oct 18 21:28:10.022 127.0.0.1 tr10000: ERROR - PushNotifications: (18/38)   at com.mendix.modules.microflowengine.microflow.impl.MicroflowImpl.executeAction(MicroflowImpl.java:149)
Oct 18 21:28:10.022 127.0.0.1 tr10000: ERROR - PushNotifications: (19/38)   at com.mendix.systemwideinterfaces.core.UserAction.execute(UserAction.java:49)
Oct 18 21:28:10.022 127.0.0.1 tr10000: ERROR - PushNotifications: (20/38)   at com.mendix.core.actionmanagement.CoreAction.doCall(CoreAction.java:260)
Oct 18 21:28:10.022 127.0.0.1 tr10000: ERROR - PushNotifications: (21/38)   at com.mendix.core.actionmanagement.CoreAction.call(CoreAction.java:248)
Oct 18 21:28:10.022 127.0.0.1 tr10000: ERROR - PushNotifications: (22/38)   at com.mendix.core.actionmanagement.CustomActionMonitorImpl$$anonfun$run$1.apply(CustomMonitoredAction.scala:14)
Oct 18 21:28:10.022 127.0.0.1 tr10000: ERROR - PushNotifications: (23/38)   at com.mendix.core.session.Worker$$anonfun$receive$3$$anonfun$2$$anon$1.execute(ActionDispatching.scala:144)
Oct 18 21:28:10.022 127.0.0.1 tr10000: ERROR - PushNotifications: (24/38)   at com.mendix.util.classloading.Runner.doRunUsingClassLoaderOf(Runner.java:36)
Oct 18 21:28:10.022 127.0.0.1 tr10000: ERROR - PushNotifications: (25/38)   at com.mendix.core.session.Worker$$anonfun$receive$3$$anonfun$2.apply(ActionDispatching.scala:146)
Oct 18 21:28:10.022 127.0.0.1 tr10000: ERROR - PushNotifications: (26/38)   at scala.util.Try$.apply(Try.scala:192)
Oct 18 21:28:10.022 127.0.0.1 tr10000: ERROR - PushNotifications: (27/38)   at com.mendix.core.session.Worker$$anonfun$receive$3.applyOrElse(ActionDispatching.scala:140)
Oct 18 21:28:10.022 127.0.0.1 tr10000: ERROR - PushNotifications: (28/38)   at akka.actor.Actor$class.aroundReceive(Actor.scala:465)
Oct 18 21:28:10.022 127.0.0.1 tr10000: ERROR - PushNotifications: (29/38)   at com.mendix.core.session.Worker.aroundReceive(ActionDispatching.scala:136)
Oct 18 21:28:10.023 127.0.0.1 tr10000: ERROR - PushNotifications: (30/38)   at akka.actor.ActorCell.receiveMessage(ActorCell.scala:516)
Oct 18 21:28:10.023 127.0.0.1 tr10000: ERROR - PushNotifications: (31/38)   at akka.actor.ActorCell.invoke(ActorCell.scala:487)
Oct 18 21:28:10.023 127.0.0.1 tr10000: ERROR - PushNotifications: (32/38)   at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:238)
Oct 18 21:28:10.023 127.0.0.1 tr10000: ERROR - PushNotifications: (33/38)   at akka.dispatch.Mailbox.run(Mailbox.scala:220)
Oct 18 21:28:10.023 127.0.0.1 tr10000: ERROR - PushNotifications: (34/38)   at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(AbstractDispatcher.scala:393)
Oct 18 21:28:10.023 127.0.0.1 tr10000: ERROR - PushNotifications: (35/38)   at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
Oct 18 21:28:10.023 127.0.0.1 tr10000: ERROR - PushNotifications: (36/38)   at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
Oct 18 21:28:10.023 127.0.0.1 tr10000: ERROR - PushNotifications: (37/38)   at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
Oct 18 21:28:10.023 127.0.0.1 tr10000: ERROR - PushNotifications: (38/38)   at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)``
```

**Comment by KevinVlaanderen** [comment]
We have replaced the Java/XMPP-based implementation of the GCM connector with a native Mendix implementation that uses the GCM HTTP/REST API. This should resolve the issue w.r.t. failing push notifications on Android due to a CONNECTION_DRAINING error. Other responses are now also properly handled (such as 'device unregistered' errors, or responses with a new canonical registration id).

You'll find the changes in PR #47. In this PR, we updated the test project. It contains an updated version of the PushNotifications module. It now has an additional dependency on the CommunityCommons module.

The demo project will be updated at a later stage.

Any feedback regarding these changes, in general or with regard to the CONNECTION_DRAINING issue, would be greatly appreciated!
